packages/lit-html/src/directives/async-replace.ts

```diff
@@ -22,57 +34,64 @@ class AsyncReplaceDirective extends AsyncDirective {
   }
 
   update(_part: ChildPart, [value, mapper]: DirectiveParameters<this>) {
+    // If our initial render occurs while disconnected, ensure that the pauser
+    // and weakThis are in the disconnected state
+    if (!this.isConnected) {
+      this.disconnected();
+    }
     // If we've already set up this particular iterable, we don't need
     // to do anything.
-    if (value === this._value) {
+    if (value === this.__value) {
       return;
     }
-    this._value = value;
-    this.__iterate(mapper);
-    return noChange;
-  }
-
-  // Separate function to avoid an iffe in update; update can't be async
-  // because its return value must be `noChange`
-  private async __iterate(mapper?: Mapper<unknown>) {
+    this.__value = value;
     let i = 0;
-    const {_value: value} = this;
-    for await (let v of value!) {
-      // Check to make sure that value is the still the current value of
-      // the part, and if not bail because a new value owns this part
-      if (this._value !== value) {
-        break;
+    const {__weakThis: weakThis, __pauser: pauser} = this;
+    // Note, the callback avoids closing over `this` so that the directive
+    // can be gc'ed before the promise resolves
+    forAwaitOf(value, async (v: unknown) => {
+      // The while loop here handles the case that the connection state
+      // thrashes, causing the pauser to resume and then get re-paused
+      while (pauser.get()) {
+        await pauser.get();
       }
+      const _this = weakThis.deref();
```

Comment:
Just for ultra clarity, suggest adding a comment here too that this avoids closing over `this` since the value is pulled out of the `weakThis` and that it's only there if the part is connected.

Reply:
Done